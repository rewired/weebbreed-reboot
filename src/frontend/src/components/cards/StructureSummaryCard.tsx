import Card from '@/components/Card';
import type { StructureSnapshot } from '@/types/simulation';

type StructureSummaryCardProps = {
  structure: StructureSnapshot;
  roomCount: number;
  zoneCount: number;
  plantCount?: number;
  averageTemperature?: number;
  averageHumidity?: number;
  averageCo2?: number;
  averagePpfd?: number;
  averageStress?: number;
  averageLightingCoverage?: number;
  isSelected?: boolean;
  onSelect?: (structureId: string) => void;
  className?: string;
};

const decimalFormatter = new Intl.NumberFormat('en-US', {
  maximumFractionDigits: 1,
});

const percentageFormatter = new Intl.NumberFormat('en-US', {
  style: 'percent',
  maximumFractionDigits: 0,
});

const currencyFormatter = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'EUR',
  maximumFractionDigits: 0,
});

const STATUS_LABEL: Record<StructureSnapshot['status'], string> = {
  active: 'Active',
  underConstruction: 'Under construction',
  decommissioned: 'Decommissioned',
};

const StructureSummaryCard = ({
  structure,
  roomCount,
  zoneCount,
  plantCount = 0,
  averageTemperature,
  averageHumidity,
  averageCo2,
  averagePpfd,
  averageStress,
  averageLightingCoverage,
  isSelected = false,
  onSelect,
  className,
}: StructureSummaryCardProps) => {
  const metadata = [
    { label: 'Footprint area', value: `${decimalFormatter.format(structure.footprint.area)} m²` },
    {
      label: 'Footprint volume',
      value: `${decimalFormatter.format(structure.footprint.volume)} m³`,
    },
  ];

  if (structure.rentPerTick > 0) {
    metadata.push({
      label: 'Rent per tick',
      value: currencyFormatter.format(structure.rentPerTick),
    });
  }

  metadata.push({ label: 'Plants', value: plantCount.toLocaleString() });

  const handleClick = () => {
    if (onSelect) {
      onSelect(structure.id);
    }
  };

  const cardClasses = [
    className,
    onSelect ? 'cursor-pointer' : undefined,
    isSelected ? 'border-accent shadow-strong ring-1 ring-accent/40' : undefined,
  ]
    .filter(Boolean)
    .join(' ');

  return (
    <Card
      title={structure.name}
      subtitle={`${roomCount.toLocaleString()} rooms • ${zoneCount.toLocaleString()} zones`}
      metadata={metadata}
      interactive={Boolean(onSelect)}
      onClick={onSelect ? handleClick : undefined}
      className={cardClasses}
    >
      <div className="space-y-3 text-sm">
        <div className="flex flex-wrap items-center gap-2 text-xs uppercase tracking-wide text-text-muted">
          <span>Status</span>
          <span className="rounded-full border border-border/50 bg-surfaceAlt/70 px-2 py-0.5 text-[0.65rem] font-semibold text-text-secondary">
            {STATUS_LABEL[structure.status] ?? structure.status}
          </span>
        </div>
        <div className="grid grid-cols-2 gap-3">
          <div>
            <p className="text-xs uppercase tracking-wide text-text-muted">Avg temperature</p>
            <p className="text-base font-medium text-text-primary">
              {averageTemperature !== undefined ? `${averageTemperature.toFixed(1)} °C` : '—'}
            </p>
          </div>
          <div>
            <p className="text-xs uppercase tracking-wide text-text-muted">Avg humidity</p>
            <p className="text-base font-medium text-text-primary">
              {averageHumidity !== undefined ? `${(averageHumidity * 100).toFixed(0)}%` : '—'}
            </p>
          </div>
          <div>
            <p className="text-xs uppercase tracking-wide text-text-muted">Avg CO₂</p>
            <p className="text-base font-medium text-text-primary">
              {averageCo2 !== undefined ? `${averageCo2.toFixed(0)} ppm` : '—'}
            </p>
          </div>
          <div>
            <p className="text-xs uppercase tracking-wide text-text-muted">Avg PPFD</p>
            <p className="text-base font-medium text-text-primary">
              {averagePpfd !== undefined ? `${averagePpfd.toFixed(0)} μmol·m⁻²·s⁻¹` : '—'}
            </p>
          </div>
          <div>
            <p className="text-xs uppercase tracking-wide text-text-muted">Avg stress</p>
            <p className="text-base font-medium text-text-primary">
              {averageStress !== undefined
                ? percentageFormatter.format(Math.max(0, Math.min(averageStress, 1)))
                : '—'}
            </p>
          </div>
          <div>
            <p className="text-xs uppercase tracking-wide text-text-muted">Lighting coverage</p>
            <p className="text-base font-medium text-text-primary">
              {averageLightingCoverage !== undefined
                ? percentageFormatter.format(Math.max(0, Math.min(averageLightingCoverage, 1)))
                : '—'}
            </p>
          </div>
        </div>
      </div>
    </Card>
  );
};

export default StructureSummaryCard;
