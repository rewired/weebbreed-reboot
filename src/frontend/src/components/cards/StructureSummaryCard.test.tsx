import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { afterEach, describe, expect, it, vi } from 'vitest';

import StructureSummaryCard from './StructureSummaryCard';
import type { StructureSnapshot } from '@/types/simulation';

const baseStructure = {
  id: 'structure-1',
  name: 'North Facility',
  status: 'active',
  footprint: { length: 40, width: 20, height: 6, area: 800, volume: 4800 },
  rentPerTick: 1250,
  roomIds: ['room-1', 'room-2'],
} satisfies StructureSnapshot;

describe('StructureSummaryCard', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders metadata, metrics and status pills', () => {
    render(
      <StructureSummaryCard
        structure={baseStructure}
        roomCount={2}
        zoneCount={4}
        plantCount={128}
        averageTemperature={24.5}
        averageHumidity={0.62}
        averageCo2={980}
        averagePpfd={540}
        averageStress={0.18}
        averageLightingCoverage={0.84}
      />,
    );

    expect(screen.getByText('North Facility')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('2 rooms • 4 zones')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('Footprint area')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('800 m²')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('Footprint volume')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('4,800 m³')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('Rent per tick')).toBeInstanceOf(HTMLElement);
    const rentMatcher = (content: string) =>
      content.replace(/\s/g, '') === '€1,250'.replace(/\s/g, '');
    expect(screen.getByText(rentMatcher)).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('Plants')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('128')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('Active')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('24.5 °C')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('62%')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('980 ppm')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('540 μmol·m⁻²·s⁻¹')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('18%')).toBeInstanceOf(HTMLElement);
    expect(screen.getByText('84%')).toBeInstanceOf(HTMLElement);
  });

  it('invokes the selection callback when clicked', () => {
    const handleSelect = vi.fn();
    const { container } = render(
      <StructureSummaryCard
        structure={baseStructure}
        roomCount={2}
        zoneCount={4}
        onSelect={handleSelect}
      />,
    );

    fireEvent.click(container.firstElementChild as HTMLElement);

    expect(handleSelect).toHaveBeenCalledWith('structure-1');
  });

  it('matches the snapshot for highlighted structures', () => {
    const { container } = render(
      <StructureSummaryCard
        structure={{ ...baseStructure, status: 'underConstruction' }}
        roomCount={0}
        zoneCount={0}
        averageStress={0.5}
        averageLightingCoverage={0.3}
        isSelected
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
