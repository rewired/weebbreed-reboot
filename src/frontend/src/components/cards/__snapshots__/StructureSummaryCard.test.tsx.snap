// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StructureSummaryCard > matches the snapshot for highlighted structures 1`] = `
<div
  class="flex h-full flex-col gap-3 rounded-lg border border-border/50 bg-surfaceAlt/70 p-4 shadow-soft transition border-accent shadow-strong ring-1 ring-accent/40"
>
  <header
    class="space-y-1"
  >
    <h3
      class="text-lg font-semibold text-text-primary"
    >
      North Facility
    </h3>
    <p
      class="text-sm text-text-secondary"
    >
      0 rooms • 0 zones
    </p>
  </header>
  <dl
    class="grid grid-cols-1 gap-3 text-xs text-text-muted sm:grid-cols-2"
  >
    <div
      class="space-y-1 rounded-md border border-border/40 bg-surfaceAlt/60 p-3"
    >
      <dt
        class="uppercase tracking-wide"
      >
        Footprint area
      </dt>
      <dd
        class="text-sm font-medium text-text-primary"
      >
        800 m²
      </dd>
    </div>
    <div
      class="space-y-1 rounded-md border border-border/40 bg-surfaceAlt/60 p-3"
    >
      <dt
        class="uppercase tracking-wide"
      >
        Footprint volume
      </dt>
      <dd
        class="text-sm font-medium text-text-primary"
      >
        4,800 m³
      </dd>
    </div>
    <div
      class="space-y-1 rounded-md border border-border/40 bg-surfaceAlt/60 p-3"
    >
      <dt
        class="uppercase tracking-wide"
      >
        Rent per tick
      </dt>
      <dd
        class="text-sm font-medium text-text-primary"
      >
        €1,250
      </dd>
    </div>
    <div
      class="space-y-1 rounded-md border border-border/40 bg-surfaceAlt/60 p-3"
    >
      <dt
        class="uppercase tracking-wide"
      >
        Plants
      </dt>
      <dd
        class="text-sm font-medium text-text-primary"
      >
        0
      </dd>
    </div>
  </dl>
  <div
    class="flex-1 text-sm text-text-secondary"
  >
    <div
      class="space-y-3 text-sm"
    >
      <div
        class="flex flex-wrap items-center gap-2 text-xs uppercase tracking-wide text-text-muted"
      >
        <span>
          Status
        </span>
        <span
          class="rounded-full border border-border/50 bg-surfaceAlt/70 px-2 py-0.5 text-[0.65rem] font-semibold text-text-secondary"
        >
          Under construction
        </span>
      </div>
      <div
        class="grid grid-cols-2 gap-3"
      >
        <div>
          <p
            class="text-xs uppercase tracking-wide text-text-muted"
          >
            Avg temperature
          </p>
          <p
            class="text-base font-medium text-text-primary"
          >
            —
          </p>
        </div>
        <div>
          <p
            class="text-xs uppercase tracking-wide text-text-muted"
          >
            Avg humidity
          </p>
          <p
            class="text-base font-medium text-text-primary"
          >
            —
          </p>
        </div>
        <div>
          <p
            class="text-xs uppercase tracking-wide text-text-muted"
          >
            Avg CO₂
          </p>
          <p
            class="text-base font-medium text-text-primary"
          >
            —
          </p>
        </div>
        <div>
          <p
            class="text-xs uppercase tracking-wide text-text-muted"
          >
            Avg PPFD
          </p>
          <p
            class="text-base font-medium text-text-primary"
          >
            —
          </p>
        </div>
        <div>
          <p
            class="text-xs uppercase tracking-wide text-text-muted"
          >
            Avg stress
          </p>
          <p
            class="text-base font-medium text-text-primary"
          >
            50%
          </p>
        </div>
        <div>
          <p
            class="text-xs uppercase tracking-wide text-text-muted"
          >
            Lighting coverage
          </p>
          <p
            class="text-base font-medium text-text-primary"
          >
            30%
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
